src/ipc/ipc-common/observer.h

```diff
@@ -0,0 +1,107 @@
+/**********************************************************************
+**
+** Copyright (C) 2020 Luxoft Sweden AB
+**
+** This file is part of the FaceLift project
+**
+** Permission is hereby granted, freIPCServiceAdapterBasee of charge, to any person
+** obtaining a copy of this software and associated documentation files
+** (the "Software"), to deal in the Software without restriction,
+** including without limitation the rights to use, copy, modify, merge,
+** publish, distribute, sublicense, and/or sell copies of the Software,
+** and to permit persons to whom the Software is furnished to do so,
+** subject to the following conditions:
+**
+** The above copyright notice and this permission notice shall be
+** included in all copies or substantial portions of the Software.
+**
+** THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+** EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+** MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+** NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+** BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+** ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+** CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+** SOFTWARE.
+**
+** SPDX-License-Identifier: MIT
+**
+**********************************************************************/
+
+#pragma once
+
+#include <memory>
+#include <QVector>
+#include <QObject>
+
+namespace facelift {
+
+class IObserver : public QObject
+{
+    Q_OBJECT
+public:
+    virtual void onReadyChanged(std::shared_ptr<QMetaObject::Connection> connection) = 0;
+};
+
+
+class IsReadyObserver: public QObject
+{
+    Q_OBJECT
+
+    QVector<IObserver *> m_observers{};
+public:
+    IsReadyObserver() {}
+
+    // Set observers
+    void setObservers(const QVector<IObserver *> &observers) {
+        m_observers = observers;
+        for(auto observer: observers){
+            auto connection = std::make_shared<QMetaObject::Connection>();
+            *connection = QObject::connect(this, &IsReadyObserver::readyChanged, observer, [observer, connection](){
+                observer->onReadyChanged( connection );
+            });
+         }
+    }
+
+    // Get observers
+    const QVector<IObserver *> &getObservers() const {
+        return m_observers;
+    }
+
+    Q_SIGNAL void readyChanged();
+
+    void onReadyChanged() {
+        emit readyChanged();
+    }
+};
+
+// Single-time observer which will unregister itself when done
+template<typename F>
```

Comment:
Use T instead - it's common type convention.

Reply:
Done